hack/verify-contextual-logging.sh

```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+
+# Copyright 2022 The KCP Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o nounset
+set -o pipefail
+# Don't set errexit because logcheck exits with non-zero when it finds errors to fix.
+
+SCRIPT_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+LOG_FILE="./${SCRIPT_ROOT}/hack/logging-results"
+
+errs=$(logcheck -check-contextual "${SCRIPT_ROOT}"/pkg/... 2>&1)
```

Comment:
I would suggest something like:

1. `set -o errexit`, always, unconditionally 
2. then opt into not exiting by checking error codes:
```sh
workfile="$( mktemp )"

function cleanup() {
    rm -rf "${workfile}"
}

trap EXIT cleanup

if logcheck -check-contextual "${SCRIPT_ROOT}"/pkg/... >"${workfile}" 2>&1; then
    echo "[ERROR] logcheck found no errors, but we expected $( wc -l < "${LOG_FILE}")"
    exit 1
fi

if [[ "${UPDATE:-}" == "true" ]]; then
    mv "${workfile}" "${LOG_FILE}"
fi

if ! changes="$( diff "${workfile}" "${LOG_FILE}" )"; then
    echo "[ERROR] <some message about changes between logcheck errors found and expected>"
    exit 1
fi
```

Reply:
I like this suggestion. Take a look at the latest revision